File: scripts/vpn_start.js
#!/usr/bin/env node

var os = require('os'),
  Q = require('q'),
  _ = require('underscore'),
  fs = require('fs'),
  dns = require('dns'),
  child_process = require('child_process');

var pia_ip,
  dnsLookup = Q.denodeify(dns.resolve4),
  writeFile = Q.denodeify(fs.writeFile),
  exec = Q.denodeify(child_process.exec);

function log(msg) {
  if (_.isArray(msg)) {
    msg = _.first(msg);
  }
  if (msg) {
    process.stdout.write(msg);
  }
}

function getIP() {
  return _.find(os.networkInterfaces(), function (iface, name) {
    return /^e(n|th)/.test(name);
  })[0].address;
}

function getNetAddr(ipRoute) {
  var ip = getIP();
  return _.find(ipRoute.split('\n'), function (line) {
    return line.indexOf(ip) !== -1;
  }).split(' ')[0];
}

function writeIpRules(stdout) {
  log('Done.\nWriting iptables rule file: ');
  return writeFile('/etc/iptables.up.rules', [
    '*filter',
    ':INPUT ACCEPT [0:0]',
    ':FORWARD ACCEPT [0:0]',
    ':OUTPUT ACCEPT [0:0]',
    '-A OUTPUT -o lo -j ACCEPT',
    '-A OUTPUT -o tun0 -j ACCEPT',
    '-A OUTPUT -d ' + getNetAddr(stdout[0]) + ' -j ACCEPT',
    '-A OUTPUT -d ' + pia_ip + ' -j ACCEPT',
    '-A OUTPUT -j DROP',
    'COMMIT' + os.EOL
  ].join(os.EOL));
}

function writeVpnConfig(addresses) {
  log('Done.\nWriting VPN config file: ');
  pia_ip = addresses[0];
  return writeFile('/etc/openvpn/client.conf', [
    'up /etc/openvpn/update-resolv-conf',
    'down /etc/openvpn/update-resolv-conf',
    'script-security 2',
    'client',
    'dev tun',
    'proto udp',
    'remote ' + pia_ip + ' 1194',
    'resolv-retry infinite',
    'nobind',
    'persist-key',
    'persist-tun',
    'ca ca.crt',
    'tls-client',
    'remote-cert-tls server',
    'auth-user-pass login.txt',
    'comp-lzo',
    'verb 1',
    'reneg-sec 0' + os.EOL
  ].join(os.EOL));
}

if (process.getuid() !== 0) {
  exec('sudo ./vpn_start.js').then(log);
} else {
  log('Getting VPN IP: ');
  dnsLookup('us-east.privateinternetaccess.com')
    .then(writeVpnConfig)
    .then(function () {
      log('Done.\nStarting OpenVPN: ');
      return exec('systemctl restart openvpn@client.service');
    })
    .then(function() {
      return exec('ip route');
    })
    .then(writeIpRules)
    .then(function () {
      log('Done.\nApplying iptables rules: ');
      return exec('iptables-restore < /etc/iptables.up.rules');
    })
    .then(function() {
      console.log('Done.');
    })
    .catch(function (err) {
       console.log('Failed.\n' + err);
    });
}

File: scripts/vpn_start.py
#! /usr/bin/python3
import os
import socket
import subprocess
import sys

def sh(cmd):
  return subprocess \
    .check_output(cmd, shell=True, universal_newlines=True) \
    .rstrip()

def writeFile(filepath, content):
  f = open(filepath, 'w')
  f.write(content)
  f.close()

def getPing(ip):
  return sh('ping -c 1 %s | grep ttl | sed -r \'s/.*time=(\S+).*/\\1/\'' % ip)

if os.geteuid() != 0:
  print('Must be run as root.')
  sys.exit()

sh('iptables -F; systemctl stop openvpn@client.service')
# find the address with the shortest ping:
results = []
for ip in socket.gethostbyname_ex('us-east.privateinternetaccess.com')[2]:
  results.append((getPing(ip), ip))
PIA = sorted(results)[0][1]

writeFile('/etc/openvpn/client.conf',
'''up /etc/openvpn/update-resolv-conf
down /etc/openvpn/update-resolv-conf
script-security 2
client
dev tun
proto udp
remote %s 1194
resolv-retry infinite
nobind
persist-key
persist-tun
ca ca.crt
tls-client
remote-cert-tls server
auth-user-pass login.txt
comp-lzo
verb 1
reneg-sec 0''' % PIA)
sh('systemctl start openvpn@client.service')

sh('''sudo iptables -A OUTPUT -o lo -j ACCEPT
sudo iptables -A OUTPUT -o tun0 -j ACCEPT
sudo iptables -A OUTPUT -d %(CIDR)s -j ACCEPT
sudo iptables -A OUTPUT -d %(PIA)s -j ACCEPT
sudo iptables -A OUTPUT -j DROP''' % {
  'CIDR': sh('ip route show scope link | grep -E "dev e(n|th)" | cut -d " " -f 1'),
  'PIA': PIA
})

print('Connected to:', PIA)
